### DISCUSSION TITLE
[BigQuery] How to instantiate request

### DISCUSSION THREAD

**Discussion by FEC-bendingspoons** [Byron/google-apis-rs#327] [Q&A] [answered]
Hello!

I am trying to use the `insert_all` method:
https://github.com/Byron/google-apis-rs/blob/beb110a941186f938996b3e928824f04f59183d3/gen/bigquery2/src/api.rs#L4841

This requires me to instantiate a `TableDataInsertAllRequest` struct:
https://github.com/Byron/google-apis-rs/blob/beb110a941186f938996b3e928824f04f59183d3/gen/bigquery2/src/api.rs#L3093

This requires me to instantiate a  vector of `TableDataInsertAllRequestRows`:
https://github.com/Byron/google-apis-rs/blob/beb110a941186f938996b3e928824f04f59183d3/gen/bigquery2/src/api.rs#L3853

This require me to provide a `JsonObject` for each row:
https://github.com/Byron/google-apis-rs/blob/beb110a941186f938996b3e928824f04f59183d3/gen/bigquery2/src/api.rs#L2035

But it has private members and no initializer that I can see. How can I create it? What am I missing?

**Comment by Byron** [comment, marked as answer]
Thanks for letting me know in such an easy to follow fashion! Apologies for turning this legitimate issue into a discussion, #328 is the re-created issue from it to which you might want to subscribe for updates.

That said, to me it appears that the tuple-structs the system generates are generally lacking the `pub` specifier and I couldn't think of any other reason than oversight.

My suggestion is to vendor the code, make the needed adjustments, and see if the valid request can be made like that. I have made the necessary adjustment to the generator and regenerated the code so you can take it from `main`. Please let me know in the issue if this works for you, and I could make a new release for bigquery2 specifically. Thanks.

**Comment by FEC-bendingspoons** [comment]
As always, it's a pleasure interacting with you.

I can confirm that the new version currently on `main` (rev: 77f097dda2e1dcb41eb40391e5a59d48ad2341b1) works 🚀 .

It is a bit unwieldy however. I have something like

```rust
#[derive(Serialize)]
struct Event {
 // ... various fields
}
```

RIght now I am keeping a buffer of them ready to be sent

```rust
let mut buffer: Vec<JsonObject> = Vec::with_capacity(CAPACITY);

loop {
  match channel.recv() {
    Some(event) => {
       // here I have to do some work
      let values: HashMap<String, JsonValue> = serde_json::to_value(event)
          .unwrap()
          .as_object()
          .unwrap()
          .into_iter()
          .map(|(key, value)| (key.clone(), JsonValue(value.clone())))
          .collect();

      buffer.push(JsonObject(Some(values)));
    }
    _ => {}
  }
  // ....
}
```

I don't have many ideas on how to make it better :(

**Comment by FEC-bendingspoons** [comment]
I played around a little bit and I think that it should be possible to add something like:
```rust
impl<T> TryFrom<T> for JsonObject where T: Serialize {
    type Error = serde_json::Error;

    fn try_from(value: T) -> Result<Self, Self::Error> {
        let values_map: HashMap<String, JsonValue> = serde_json::to_value(value)?
            .as_object()
            .ok_or_else(|| {
                serde_json::Error::custom("value needs to be representable as a JSON object")
            })?
            .into_iter()
            .map(|(key, value)| (key.clone(), JsonValue(value.clone())))
            .collect();

        Ok(JsonObject(Some(values_map)))
    }
}
```
but the compiler complains about a duplicated conformance, and I cannot figure out why 😢

**Reply by Byron** [reply]
Thank you :). And you are welcome.

And even if the above would work, the problem is that `JsonObject` is fully generated and it would be difficult/complicated to add implementations for it.

Since it's now working I would take this home as a win and hope that with utility functions it should be possible to ease the pain of using it.

**Reply by FEC-bendingspoons** [reply]
Yea, I share the same feeling. Thanks for the support!

**Comment by FEC-bendingspoons** [comment]
@Byron 

I think something went wrong somewhere with the release... I can see that the regeneration commit (77f097dda2e1dcb41eb40391e5a59d48ad2341b1) bumped the version to `3.0.1` 6 days ago, but according to [crates.io](https://crates.io/crates/google-bigquery2) version `3.0.1` was released 25 days ago.

Am I missing something, or are things really broken?

**Reply by Byron** [reply]
Sorry for the hassle, this is quite confusion indeed and I don't know what happened there. So I just fixed it by releasing v3.0.2, which _should_ be the version you are looking for.

**Reply by FEC-bendingspoons** [reply]
Amazing, thank you!
